client/trino-client/src/main/java/io/trino/client/spooling/encoding/DecompressionUtils.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.client.spooling.encoding;
+
+import io.airlift.compress.lz4.Lz4Decompressor;
+import io.airlift.compress.zstd.ZstdDecompressor;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.Optional;
+
+public class DecompressionUtils
+{
+    private static final MethodHandles.Lookup LOOKUP = MethodHandles.lookup();
+    private static final Optional<MethodHandle> LZ4_DECOMPRESSOR_CONSTRUCTOR =
+            getDecompressorConstructor("io.airlift.compress.v3.lz4.Lz4NativeDecompressor");
+    private static final Optional<MethodHandle> ZSTD_DECOMPRESSOR_CONSTRUCTOR =
+            getDecompressorConstructor("io.airlift.compress.v3.zstd.ZstdNativeDecompressor");
+
+    private DecompressionUtils() {}
+
+    public static int decompressLZ4(byte[] input, byte[] output)
+    {
+        if (!LZ4_DECOMPRESSOR_CONSTRUCTOR.isPresent()) {
```

Comment:
should you check for exceptions here also, and fallback to non-native on next try?